webui/react/src/components/FilterForm/components/FilterFormStore.ts

```diff
@@ -0,0 +1,227 @@
+import { observable, Observable, WritableObservable } from 'micro-observables';
+
+import {
+  AvailableOperators,
+  Conjunction,
+  FilterFormSet,
+  FilterFormSetWithoutId,
+  FormField,
+  FormFieldValue,
+  FormGroup,
+  FormKind,
+  Operator,
+} from 'components/FilterForm/components/type';
+import { V1ColumnType, V1LocationType, V1ProjectColumn } from 'services/api-ts-sdk';
+
+export const ITEM_LIMIT = 50;
+
+export const ROOT_ID = 'ROOT';
+
+export const INIT_FORMSET: Readonly<FilterFormSet> = {
+  filterGroup: { children: [], conjunction: Conjunction.And, id: ROOT_ID, kind: FormKind.Group },
+  showArchived: false,
+};
+
+const getInitGroup = (): FormGroup => ({
+  children: [],
+  conjunction: Conjunction.And,
+  id: crypto.randomUUID(),
+  kind: FormKind.Group,
+});
+
+const getInitField = (): FormField => ({
+  columnName: 'name',
+  id: crypto.randomUUID(),
+  kind: FormKind.Field,
+  location: V1LocationType.EXPERIMENT,
+  operator: AvailableOperators[V1ColumnType.TEXT][0],
+  type: V1ColumnType.TEXT,
+  value: null,
+});
+
+export class FilterFormStore {
+  #formset: WritableObservable<FilterFormSet> = observable(structuredClone(INIT_FORMSET));
+
+  public init(data?: Readonly<FilterFormSet>): void {
+    this.#formset.update(() => structuredClone(data ? data : INIT_FORMSET));
+  }
+
+  public get formset(): Observable<Readonly<FilterFormSet>> {
+    return this.#formset.readOnly();
+  }
+
+  public get jsonWithoutId(): Observable<Readonly<FilterFormSetWithoutId>> {
+    const replacer = (key: string, value: unknown): unknown => {
+      return key === 'id' ? undefined : value;
+    };
+    return this.#formset.select((formset) => {
+      const newFormSet: FilterFormSetWithoutId = JSON.parse(JSON.stringify(formset, replacer));
+      return newFormSet;
+    });
+  }
+
+  public get fieldCount(): number {
+    const countFields = (form: FormGroup): number => {
+      let count = 0;
+      for (const child of form.children) {
+        count += child.kind === FormKind.Group ? countFields(child) : 1;
+      }
+      return count;
+    };
+    return countFields(this.#formset.get().filterGroup);
+  }
```

Comment:
This is read on every render, but should only change when the formset changes. Consider selecting this instead:
```suggestion
  public get fieldCount(): Observable<number> {
    const countFields = (form: FormGroup): number => {
      let count = 0;
      for (const child of form.children) {
        count += child.kind === FormKind.Group ? countFields(child) : 1;
      }
      return count;
    };
    return this.#formset.select(({ filterGroup }) => countFields(filterGroup));
  }
```